File: IdeaCards/Classes/Controller/EditController.m
#import "EditController.h"
#import "AppDelegate.h"
#import "TopController.h"
#import "ICItemManager.h"

@interface EditController ()
<UITableViewDataSource, UITableViewDelegate, UITextFieldDelegate>

@property(nonatomic,retain) UITableView *table;
@property(nonatomic,retain) UITextField *addField;
@property(nonatomic,retain) UIButton *addButton;
@property(nonatomic,retain) UIView *addView;

@end

@implementation EditController


#pragma mark - View Lifecycle

- (void)loadView
{
    [super loadView];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    // Table View
    CGRect frame;
    if(self.type == EditTypeRight){
        if((self.interfaceOrientation == UIInterfaceOrientationLandscapeLeft ||
            self.interfaceOrientation == UIInterfaceOrientationLandscapeRight)){
            frame = CGRectMake(30,
                               0,
                               self.view.frame.size.height-30,
                               self.view.frame.size.width);
            
        }else{
            frame = CGRectMake(30,
                               0,
                               self.view.frame.size.width-30,
                               self.view.frame.size.height);
            
        }
        
    }else{
        if((self.interfaceOrientation == UIInterfaceOrientationLandscapeLeft ||
            self.interfaceOrientation == UIInterfaceOrientationLandscapeRight)){
            frame = CGRectMake(0,
                               0,
                               self.view.frame.size.height-30,
                               self.view.frame.size.width);
            
        }else{
            frame = CGRectMake(0,
                               0,
                               self.view.frame.size.width-30,
                               self.view.frame.size.height);
            
        }
    }
    
    self.table = [[UITableView alloc] initWithFrame:frame
                                              style:UITableViewStylePlain];
    self.table.delegate = self;
    self.table.dataSource = self;
    [self.view addSubview:self.table];
    
    
    // Input
    UIView *buttonView = [UIView new];
    buttonView.frame = CGRectMake(0, 0, self.view.frame.size.width, 44);
    buttonView.backgroundColor = [UIColor clearColor];
    
    self.addField = [UITextField new];
    if(self.type == EditTypeRight){
        self.addField.frame = CGRectMake(40, 5, 220, 34);
    }else{
        self.addField.frame = CGRectMake(0, 5, 220, 34);
    }
    self.addField.delegate = self;
    self.addField.borderStyle = UITextBorderStyleRoundedRect;
    self.addField.returnKeyType = UIReturnKeyDone;
    [buttonView addSubview:self.addField];
    
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    button.frame = CGRectMake(self.addField.frame.origin.x + self.addField.frame.size.width, 0, 44, 44);
    button.titleLabel.font = [UIFont fontWithName:@"Helvetica-Bold" size:32.0f];
    button.titleEdgeInsets = UIEdgeInsetsMake(-5.0f, 0.0f, 0.0f, 0.0f);
    [button setTitle:@"+" forState:UIControlStateNormal];
    [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [button setBackgroundColor:[UIColor clearColor]];
    button.titleLabel.shadowColor = [UIColor blackColor];
    button.titleLabel.shadowOffset = CGSizeMake(0.0f, -1.0f);
    [button addTarget:self action:@selector(onAddButton:) forControlEvents:UIControlEventTouchUpInside];
    [buttonView addSubview:button];
    self.addButton = button;
    
    UIBarButtonItem *addItem = [[UIBarButtonItem alloc] initWithCustomView:buttonView];
    self.navigationItem.leftBarButtonItem = addItem;
    
    self.addView = buttonView;
    
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(orientationChanged:)
                                                 name:UIDeviceOrientationDidChangeNotification
                                               object:nil];
    
    [self orientationChanged:nil];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIDeviceOrientationDidChangeNotification
                                                  object:nil];
}


#pragma mark - Action

- (void)onAddButton:(UIBarButtonItem *)button
{
    [self.addField resignFirstResponder];
    
    [self _add];
}


#pragma mark - Notification

- (void)orientationChanged:(NSNotification *)notification
{
    if(self.interfaceOrientation == UIInterfaceOrientationPortrait ||
       self.interfaceOrientation == UIInterfaceOrientationPortraitUpsideDown){
        if(self.type == EditTypeRight){
            self.table.frame = CGRectMake(30,
                                          0,
                                          320-30,
                                          480);
            
            self.addField.frame = CGRectMake(40, 5, 220, 34);
            
        }else{
            self.table.frame = CGRectMake(0,
                                          0,
                                          320-30,
                                          480);
            
            self.addField.frame = CGRectMake(0, 5, 220, 34);
        }
        
        self.addButton.frame = CGRectMake(self.addField.frame.origin.x + self.addField.frame.size.width, 0, 44, 44);
        self.addView.frame = CGRectMake(0, 0, 320, 44);
        
    }else{
        
        if(self.type == EditTypeRight){
            self.table.frame = CGRectMake(48,
                                          0,
                                          480-48,
                                          320);
            
            self.addField.frame = CGRectMake(48, 8, 350, 26);
            
        }else{
            self.table.frame = CGRectMake(0,
                                          0,
                                          480-48,
                                          320);
            
            self.addField.frame = CGRectMake(0, 8, 350, 26);
            
        }
        
        self.addButton.frame = CGRectMake(self.addField.frame.origin.x + self.addField.frame.size.width, 0, 44, 44);
        self.addView.frame = CGRectMake(0, 0, 480, 44);
        
    }
}


#pragma mark - UITableView DataSource/Delegate

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if(self.type == EditTypeRight){
        return [[[ICItemManager sharedManager] rightItems] count];
    }else{
        return [[[ICItemManager sharedManager] leftItems] count];
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"CellIdentifier";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if(cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault
                                      reuseIdentifier:CellIdentifier];
    }
    
    
    NSArray *items;
    if(self.type == EditTypeRight){
        items = [[ICItemManager sharedManager] rightItems];
    }else{
        items = [[ICItemManager sharedManager] leftItems];
    }
    
    cell.textLabel.text = [items objectAtIndex:indexPath.row];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if(self.type == EditTypeRight){
        [[ICItemManager sharedManager] deleteRightItemAtIndex:indexPath.row];
    }else{
        [[ICItemManager sharedManager] deleteLeftItemAtIndex:indexPath.row];
    }
    
    [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                     withRowAnimation:UITableViewRowAnimationFade];
    
    [self _reloadData];
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleDelete;
}


#pragma mark - UITextField Delegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [textField resignFirstResponder];
    
    [self _add];
    
    return YES;
}


#pragma mark -

- (void)_reloadData
{
    [self.table reloadData];
    
    // メイン画面のデータも更新
    AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
    if(appDelegate){
        UINavigationController *navController = (UINavigationController *)appDelegate.basePanelController.centerPanel;
        TopController *topController = (TopController *)[navController.viewControllers objectAtIndex:0];
        
        if([topController respondsToSelector:@selector(reloadData)]){
            [topController reloadData];
        }
    }
}

- (void)_add
{
    if([self.addField.text length] <= 0){
        return;
    }
    
    if(self.type == EditTypeRight){
        [[ICItemManager sharedManager] addRightItem:self.addField.text];
    }else{
        [[ICItemManager sharedManager] addLeftItem:self.addField.text];
    }
    
    self.addField.text = nil;
    
    [self _reloadData];
}

@end

File: IdeaCards/Classes/Controller/EditController.swift
//
//  EditController.swift
//  IdeaCards
//
//  Created by Kenta on 2017/03/02.
//  Copyright © 2017年 UROURO. All rights reserved.
//

import UIKit

enum EditType {
    case left
    case right
}

class EditController: UIViewController {
    
    var type: EditType = .left
    
    fileprivate lazy var tableView: UITableView = {
        let v = UITableView(frame: CGRect.zero, style: .plain)
        v.delegate = self
        v.dataSource = self
        
        switch self.type {
        case .left:
            v.contentInset = UIEdgeInsets(top: 0.0, left: 0.0, bottom: 0.0, right: 0.0)
        case .right:
            v.contentInset = UIEdgeInsets(top: 0.0, left: 0.0, bottom: 0.0, right: 0.0)
        }

        
        return v
    }()
    
    fileprivate lazy var addView: UIView = {
        let v = UIView(frame: CGRect(x: 0.0,
                                     y: 0.0,
                                     width: self.view.frame.size.width,
                                     height: 44.0))
        v.backgroundColor = UIColor.clear
        
        v.addSubview(self.addField)
        v.addSubview(self.addButton)
        
        return v
    }()
    
    fileprivate lazy var addField: UITextField = {
        let field = UITextField()
        switch self.type {
        case .right:
            field.frame = CGRect(x: 40.0, y: 5.0, width: 220.0, height: 34.0)
        case .left:
            field.frame = CGRect(x: 0.0, y: 5.0, width: 220.0, height: 34.0)
        }
        field.delegate = self
        field.borderStyle = .roundedRect
        field.returnKeyType = .done
        field.placeholder = "キーワードを入力"
        return field
    }()
    
    fileprivate lazy var addButton: UIButton = {
        let button = UIButton(type: .custom)
        button.frame = CGRect(x: self.addField.frame.origin.x + self.addField.frame.size.width,
                              y: 0.0,
                              width: 44.0,
                              height: 44.0)
        button.titleLabel?.font = UIFont(name: "Helvetica-Bold", size: 32.0)
        button.titleEdgeInsets = UIEdgeInsets(top: -5.0, left: 0.0, bottom: 0.0, right: 0.0)
        button.setTitle("+", for: .normal)
        button.setTitleColor(UIColor.darkGray, for: .normal)
        button.backgroundColor = UIColor.clear
        button.titleLabel?.shadowColor = UIColor.black
        button.titleLabel?.shadowOffset = CGSize(width: 0.0, height: -1.0)
        button.addTarget(self, action: #selector(EditController.onAddButton(_:)), for: .touchUpInside)
        return button
    }()
    
}

// MARK: - View Lifecycle
extension EditController {
    
    override func loadView() {
        super.loadView()
        
        view.backgroundColor = UIColor.white
        
        view.addSubview(tableView)
        
        navigationItem.leftBarButtonItem = UIBarButtonItem(customView: addView)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        NotificationCenter.default.addObserver(self, selector: #selector(EditController.orientationChanged(_:)), name: Notification.Name.UIDeviceOrientationDidChange, object: nil)
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        orientationChanged(nil)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        NotificationCenter.default.removeObserver(self, name: Notification.Name.UIDeviceOrientationDidChange, object: nil)
    }
    
}


// MARK: - Action
extension EditController {
    
    func onAddButton(_ sender: UIButton) {
        addField.resignFirstResponder()
        
        add()
    }
    
    func orientationChanged(_ notification: Notification?) {
        let orientation = UIApplication.shared.statusBarOrientation
        
        switch orientation {
        case .portrait, .portraitUpsideDown, .unknown:
            switch type {
            case .right:
                tableView.frame = CGRect(x: 40.0,
                                         y: 0.0,
                                         width: view.frame.size.width - 40.0,
                                         height: view.frame.size.height)
                addField.frame = CGRect(x: 56.0,
                                        y: 5.0,
                                        width: 220.0,
                                        height: 34.0)
            case .left:
                tableView.frame = CGRect(x: 0.0,
                                         y: 0.0,
                                         width: view.frame.size.width - 40.0,
                                         height: view.frame.size.height)
                addField.frame = CGRect(x: 16.0,
                                        y: 5.0,
                                        width: 220.0,
                                        height: 34.0)
            }
            
            addButton.frame = CGRect(x: addField.frame.origin.x + addField.frame.size.width,
                                     y: 0.0,
                                     width: 44.0,
                                     height: 44.0)
            addView.frame = CGRect(x: 0.0,
                                   y: 0.0,
                                   width: view.frame.size.width,
                                   height: 44.0)
            
        case .landscapeLeft, .landscapeRight:
            switch type {
            case .right:
                tableView.frame = CGRect(x: 48.0,
                                         y: 0.0,
                                         width: view.frame.size.height - 48.0,
                                         height: view.frame.size.width)
                addField.frame = CGRect(x: 48.0,
                                        y: 8.0,
                                        width: view.frame.size.width + 30.0,
                                        height: 26.0)
                
            case .left:
                tableView.frame = CGRect(x: 0.0,
                                         y: 0.0,
                                         width: view.frame.size.height - 48.0,
                                         height: view.frame.size.width)
                addField.frame = CGRect(x: 0.0,
                                        y: 8.0,
                                        width: view.frame.size.width + 30.0,
                                        height: 26.0)
            }
            
            addButton.frame = CGRect(x: addField.frame.origin.x + addField.frame.size.width,
                                     y: 0.0,
                                     width: 44.0,
                                     height: 44.0)
            addView.frame = CGRect(x: 0.0,
                                   y: 0.0,
                                   width: view.frame.size.height,
                                   height: 44.0)
        }
    }
    
}

extension EditController: UITableViewDataSource {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch type {
        case .right:
            return ICItemManager.shared().rightItems().count
        case .left:
            return ICItemManager.shared().leftItems().count
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let identifier = "CellIdentifier"
        var cell = tableView.dequeueReusableCell(withIdentifier: identifier)
        
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: identifier)
        }
        
        let items: [Any]
        
        switch type {
        case .right:
            items = ICItemManager.shared().rightItems()
        case .left:
            items = ICItemManager.shared().leftItems()
        }
        
        cell!.textLabel?.text = items[indexPath.row] as? String
        return cell!
    }
    
}

extension EditController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        switch type {
        case .right:
            ICItemManager.shared().deleteRightItem(at: indexPath.row)
        case .left:
            ICItemManager.shared().deleteLeftItem(at: indexPath.row)
        }
        
        tableView.deleteRows(at: [indexPath], with: .fade)
        
        reloadData()
    }
    
    func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        return .delete
    }
    
}

extension EditController: UITextFieldDelegate {
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        
        add()
        
        return true
    }
    
}

fileprivate extension EditController {
    
    func reloadData() {
        tableView.reloadData()
        
        let appDelegate: AppDelegate = UIApplication.shared.delegate as! AppDelegate
        
        if let basePanelController = appDelegate.basePanelController {
            let navController: UINavigationController = basePanelController.centerPanel as! UINavigationController
            let topController: TopController = navController.viewControllers[0] as! TopController
            topController.reloadData()
        }
    }
    
    func add() {
        guard let text = addField.text else {
            return
        }
        
        if text.characters.count <= 0 {
            return
        }
        
        switch type {
        case .right:
            ICItemManager.shared().addRightItem(text)
        case .left:
            ICItemManager.shared().addLeftItem(text)
        }
        
        addField.text = nil
        
        reloadData()
    }
    
}
